Adds preliminary support of transactions

rystsov opened the pull request:
Description TBD

dotnwat commented:
@rystsov 

```
../src/v/kafka/protocol/add_offsets_to_txn.h:16:10: fatal error: kafka/protocol/schemata/add_offsets_to_txn_request.h: No such file or directory
17738
   16 | #include "kafka/protocol/schemata/add_offsets_to_txn_request.h"
17739
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

this is probably a race condition because libcluster does not depend on libkafka. that file is code generated in libkafka, but doesn't guarantee that libkafka should be built before libcluster.

senior7515 commented:
> @rystsov
> 
> ```
> ../src/v/kafka/protocol/add_offsets_to_txn.h:16:10: fatal error: kafka/protocol/schemata/add_offsets_to_txn_request.h: No such file or directory
> 17738
>    16 | #include "kafka/protocol/schemata/add_offsets_to_txn_request.h"
> 17739
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> this is probably a race condition because libcluster does not depend on libkafka. that file is code generated in libkafka, but cmake doesn't guarantee that libkafka should be built before libcluster.

Typically we solve these by moving types to shared folders like model

dotnwat commented:
> > @rystsov
> > ```
> > ../src/v/kafka/protocol/add_offsets_to_txn.h:16:10: fatal error: kafka/protocol/schemata/add_offsets_to_txn_request.h: No such file or directory
> > 17738
> >    16 | #include "kafka/protocol/schemata/add_offsets_to_txn_request.h"
> > 17739
> >       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > ```
> > 
> > 
> > this is probably a race condition because libcluster does not depend on libkafka. that file is code generated in libkafka, but cmake doesn't guarantee that libkafka should be built before libcluster.
> 
> Typically we solve these by moving types to shared folders like model

This is a good suggestion. I think in this case you'll probably want to create an intermediate type and translate the request/response across kafka::* and cluster::* boundaries. Another option would be to have cluster depend on libv_kafka_protocol which is a helper library that only contains the code generated bits. Ultimately it would be nice to decide if these sub-systems in cluster:: that have kafka:: dependencies should be moved to kafka:: the answer probably depends a bit on what the future plans are.

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -246,14 +246,14 @@ configuration::configuration()
       "failing a request",
       required::no,
-      10)
+      30)
```
add a note to the commit message explaining why, for retries/delay

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -246,14 +246,14 @@ configuration::configuration()
       "Delay before retry a topic lookup in a shard or other meta tables",
       required::no,
-      0'100ms)
+      0'500ms)
```
with more retries, should we favor shorter delay to keep responsiveness low?

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+
+    auto retries = _metadata_dissemination_retries;
+    auto delay_ms = _metadata_dissemination_retry_delay_ms;
+    std::optional<model::node_id> leader_opt = std::nullopt;
```
nit: std::nullopt is already the default value.

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+    std::optional<model::node_id> leader_opt = std::nullopt;
+    tx_errc ec;
+    while (!leader_opt && 0 < retries--) {
+        ntp_opt = _coordinator_mapper.local().ntp_for(group_id);
+        if (!ntp_opt) {
+            vlog(clusterlog.trace, "can't find ntp for {}, retrying", group_id);
+            ec = tx_errc::partition_not_exists;
+            co_await ss::sleep(delay_ms);
+            continue;
+        }
```
is this needed? it looks like `ntp_opt` will always have a value given what happens above at the beginning of the method.

rystsov replied:
not really `try_create_consumer_group_topic` fills `_coordinator_mapper` but it doesn't happen instantly so there is a lag after it yields control flow but before `_coordinator_mapper` is populated

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+        }
+
+        auto ntp = ntp_opt.value();
+        auto nt = model::topic_namespace(ntp.ns, ntp.tp.topic);
```
there may be a `model::topic_namespace_view` type that avoids copies

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+        vlog(clusterlog.warn, "can't find ntp for {} ", group_id);
+        co_return prepare_group_tx_reply{.ec = tx_errc::partition_not_exists};
+    }
+    auto ntp = ntp_opt.value();
```
`std::move(ntp_opt.value())`?

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+        ss::this_shard_id(),
+        leader,
+        timeout,
+        [group_id, pid, timeout](tx_gateway_client_protocol cp) {
```
keep in mind opportunities for moving. here group_id is copied into the capture, and then copied again below into the request. in both cases it looks like it could be moved. some of the group-ids I've seen from clients have been pretty big strings some prefix and uuids etc... I think this applies to the other dispatch methods too.

dotnwat commented on src/v/cluster/rm_group_frontend.cc:
```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+    co_return co_await _group_router.local().abort_tx(std::move(req));
+}
+
+ss::future<bool> rm_group_frontend::try_create_consumer_group_topic() {
```
seems like we should pull this out into a common location and share it with find_coordinator.cc which I think does the same thing.

dotnwat commented on src/v/cluster/tx_gateway.cc:
```diff
@@ -56,24 +57,32 @@ tx_gateway::abort_tx(abort_tx_request&&, rpc::streaming_context&) {
-    co_return prepare_group_tx_reply();
+  prepare_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_frontend.local().do_prepare_group_tx(
+      request.group_id,
```
`std::move(request.group_id)`

dotnwat commented on src/v/cluster/tx_gateway.cc:
```diff
@@ -56,24 +57,32 @@ tx_gateway::abort_tx(abort_tx_request&&, rpc::streaming_context&) {
-    co_return begin_group_tx_reply();
+ss::future<begin_group_tx_reply> tx_gateway::begin_group_tx(
+  begin_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_frontend.local().do_begin_group_tx(
```
did you mean to call `prepare_group_tx` which would repeat the check of the local node being a leader and re-dispatch locally or remotely if needed. this here looks like it assumes we're on the correct node and calls the local group router directly. same question for the rest of these rpc handlers.

rystsov replied:
No, I intented to call `do_` method. `tx_gateway_frontend` invokes `rm_group_frontend::prepare_group_tx` which may dispatch a call to a remote node using `tx_gateway` so when it processes a request it uses `do_prepare_group_tx` because it already should be on the right node.

If `tx_gateway` uses `prepare_group_tx` it may create a infinite bouncing rpc loop when metadata happen to be stale.

dotnwat commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -51,4 +51,413 @@ rm_partition_frontend::rm_partition_frontend(
+  model::ntp ntp,
+  model::producer_identity pid,
+  model::timeout_clock::duration timeout) {
+    auto nt = model::topic_namespace(ntp.ns, ntp.tp.topic);
```
use `model::topic_namespace_view`?

dotnwat commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -51,4 +51,413 @@ rm_partition_frontend::rm_partition_frontend(
+          begin_tx_reply{.ntp = ntp, .ec = tx_errc::partition_not_exists});
+    }
+
+    auto leader = _leaders.local().get_leader(ntp);
```
Noah's placeholder. Will continue tomorrow

dotnwat reviewed (commented):
looking great :) I didn't make it all the way through yet but I'll finish tomorrow.

rystsov commented on src/v/cluster/rm_stm.h:
```diff
@@ -153,7 +153,7 @@ class rm_stm final : public persisted_stm {
         absl::btree_set<model::offset> ongoing_set;
         absl::flat_hash_map<model::producer_identity, prepare_marker> prepared;
-        absl::flat_hash_map<model::producer_identity, tx_range> aborted;
+        std::vector<tx_range> aborted;
```
you're right, it's also problematic because this vector is reconstructed on restart so it will eventually cause OOM, I've already have a task to this https://app.clubhouse.io/vectorized/story/2204

dotnwat commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+          _rm_partition_frontend.local().abort_tx(rm.ntp, tx.pid, timeout));
+    }
+    std::vector<ss::future<abort_group_tx_reply>> gfs;
+    for (auto group : tx.groups) {
```
this makes two copies of the group: one for the iteration object then one passed to abort_group_tx

dotnwat commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+    auto prs = co_await when_all_succeed(pfs.begin(), pfs.end());
+    auto grs = co_await when_all_succeed(gfs.begin(), gfs.end());
+    bool ok = true;
+    for (auto r : prs) {
```
`for (const auto& r : prs)`

dotnwat commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    return stm->get_end_lock(tx.id)->with([this, &stm, timeout, tx, outcome]() {
```
why capture a shared pointer by reference in the lambda? take stm by value as a parameter then move it into the capture.

rystsov replied:
I don't remember, despite being the last PR I wrote transactions using top down approach and  `tx_gateway_frontend` was among the first changes, fixing it

dotnwat commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+    std::vector<ss::future<prepare_tx_reply>> pfs;
+    for (auto rm : tx.partitions) {
+        pfs.push_back(_rm_partition_frontend.local().prepare_tx(
+          rm.ntp,
```
since `tx` is taken by value, all of its non-trivially-copyable members can be moved into prepare_tx, and `rm` can be a reference.

dotnwat commented on src/v/cluster/tx_gateway_frontend.h:
```diff
@@ -26,7 +30,15 @@
+    kafka::error_code error;
+    model::node_id node;
+    ss::sstring host;
+    int32_t port;
```
how are host/port used? we have structures like `unresolved_address` that store a host and port. seastar also has some types that might be appropriate depending on the use case.

rystsov replied:
it's a rudiment, this structure is never used, removing it altogether

dotnwat commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -354,9 +355,10 @@ produce_handler::handle(request_context ctx, ss::smp_service_group ssg) {
-        // will be removed when all of transactions are implemented
-        return ctx.respond(request.make_error_response(
-          error_code::transactional_id_authorization_failed));
+        if (!config::shard_local_cfg().enable_transactions.value()) {
```
maybe set a flag `ctx.server().enable_transactions` on the server to cache this config value?

dotnwat commented on src/v/kafka/server/handlers/find_coordinator.h:
```diff
@@ -14,6 +14,6 @@
 namespace kafka {
 
-using find_coordinator_handler = handler<find_coordinator_api, 0, 2>;
+using find_coordinator_handler = handler<find_coordinator_api, 0, 3>;
```
why is this being changed? we don't yet support the flexible protocol

rystsov replied:
it's an oversight, keeping the old version

dotnwat commented on src/v/kafka/protocol/fetch.h:
```diff
@@ -488,10 +488,16 @@ struct op_context {
 
 struct fetch_config {
```
please describe in the commit message high level descriptions of what is changing to fetch handling.

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -454,17 +471,46 @@ ss::future<read_result> read_from_ntp(
                                                         : high_watermark;
+
+    if (config::shard_local_cfg().enable_transactions.value()) {
+        if (config.isolation_level == 1) {
```
isolation level should be a strong type and the levels can have constexpr names predefined?

dotnwat reviewed (commented):
looks really good. most of my feedback is clean-up. some high level comments:

1. be careful with your log level. at the default info level warn/error messages will be logged. some of the warnings don't seem to be useful to log at this level (e.g. a timeout or other conditions that occasionally happen but resolve themselves). trace/debug are more for developers.

2. there are a lot of opportunities to reduce copies by moving. I didn't mark very many of the cases. I'm happy to go back and mark them if that's useful, but I think you'd find a lot of the cases by reviewing your code with that in mind.

3. seems reasonable to create a struct that you own and control at the cluster::* level instead of passing around the kafka::xyz_request_data etc... that will break the dependency that's causing the build issues.

4. please add details to more of your commit messages. in particular, changes in the kafka layer like the changes to fetch. please explain what is happening with the goal of giving someone intuition about the change. I promise we will want to have that context in the future.

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+    auto i = 0;
+    while (!leader_opt && i < 30) {
+        co_await ss::sleep(0'500ms);
+        leader_opt = _leaders.local().get_leader(model::kafka_tx_ntp);
```
good point! also I should use configuration instead of magic numbers

rystsov commented:
> > > @rystsov
> > > ```
> > > ../src/v/kafka/protocol/add_offsets_to_txn.h:16:10: fatal error: kafka/protocol/schemata/add_offsets_to_txn_request.h: No such file or directory
> > > 17738
> > >    16 | #include "kafka/protocol/schemata/add_offsets_to_txn_request.h"
> > > 17739
> > >       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > > ```
> > > 
> > > 
> > > this is probably a race condition because libcluster does not depend on libkafka. that file is code generated in libkafka, but cmake doesn't guarantee that libkafka should be built before libcluster.
> > 
> > 
> > Typically we solve these by moving types to shared folders like model
> 
> This is a good suggestion. I think in this case you'll probably want to create an intermediate type and translate the request/response across kafka::* and cluster::* boundaries.

@dotnwat @senior7515  Thanks for the suggestions, I've introduced intermediate type to pass the requests. Also there was another similar problem: `rm_group_frontend` depends on `group_router` which itself depends on auto-generated types. I moved `rm_group_frontend` to `kafka/server` to other consumer group code to fix it

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -49,6 +49,7 @@ static model::record_batch make_control_batch(
       raft::data_batch_type, model::offset(0));
     builder.set_producer_identity(pid.id, pid.epoch);
     builder.set_control_type();
+    builder.set_transactional_type();
```
renamed to make_tx_control_batch

rystsov commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -51,4 +51,413 @@ rm_partition_frontend::rm_partition_frontend(
+                begin_tx_reply{.ntp = ntp, .ec = tx_errc::stm_not_found});
+          }
+
+          return stm->begin_tx(pid).then([ntp](auto etag) {
```
got rid of auto there

rystsov commented on src/v/cluster/types.h:
```diff
@@ -106,14 +106,44 @@ inline std::error_code make_error_code(tx_errc e) noexcept {
+    model::producer_identity pid;
+    model::timeout_clock::duration timeout;
+};
+struct abort_tx_reply {
```
in theory yes but on practice we don't do versioning neither for `raftgen.json` nor for `metadata_dissemination_rpc.json` so I don't want to be the first to pay this price, let's fix it later for every internal api

dotnwat replied:
will any of these messages change with the follow up txn work?

rystsov replied:
maybe slightly I expect to stabilize them before they're ready for prod

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -13,14 +13,38 @@
+using namespace std::chrono_literals;
+
+static kafka::add_partitions_to_txn_response_data
+make_add_partitions_error_response(
```
I got rid of using kafka auto generated request types in cluster:: altogether

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+          tx.status == tm_transaction::tx_status::finished,
+          "unexpected tx status {}",
+          tx.status);
+    }
```
what do you mean?

dotnwat replied:
it should be deleted. it was so I could find my spot in review

rystsov commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -51,4 +51,413 @@ rm_partition_frontend::rm_partition_frontend(
+      *shard,
+      _ssg,
+      [pid, ntp, tx_seq, timeout](cluster::partition_manager& mgr) mutable {
+          vlog(clusterlog.warn, "timeout {}", timeout);
```
oops, it's a leftover from a print debug session, removing

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+                      res_partition.error_code
+                        = kafka::error_code::unknown_server_error;
+                  }
+                  topic_it->results.push_back(res_partition);
```
the assertions above covers the current case too so we don't need to assert the same thing again

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+    init_tm_tx_reply reply;
+    if (tx.pid.epoch < std::numeric_limits<int16_t>::max()) {
+        reply.pid = model::producer_identity{
+          .id = tx.pid.id, .epoch = static_cast<int16_t>(tx.pid.epoch + 1)};
```
right but `1` isn't

rystsov commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -0,0 +1,54 @@
+  ss::sharded<cluster::metadata_cache>& metadata_cache,
+  ss::sharded<rpc::connection_cache>& connection_cache,
+  ss::sharded<partition_leaders_table>& leaders,
+  std::unique_ptr<cluster::controller>& controller,
```
it's taken care by  `construct_service` and `_deferred` machinery

rystsov commented on src/v/kafka/server/handlers/add_offsets_to_txn.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace kafka {
+
+using add_offsets_to_txn_handler = handler<add_offsets_to_txn_api, 0, 2>;
```
thanks! set to one and checked other handlers

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+
+          auto outcome = ss::make_lw_shared<ss::promise<tx_errc>>();
+          if (request.committed) {
+              (void)self.commit_tm_tx(stm, tx, timeout, outcome);
```
wrapped in a gate

rystsov commented on src/v/cluster/partition.cc:
```diff
@@ -21,13 +21,25 @@ static bool is_id_allocator_topic(model::ntp ntp) {
            && ntp.tp.topic == model::id_allocator_topic;
 }
 
+static bool is_tm_topic(model::ntp ntp) {
```
yeah, bad naming, renamed

rystsov commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -51,4 +51,413 @@ rm_partition_frontend::rm_partition_frontend(
+    auto shard = _shard_table.local().shard_for(ntp);
+
+    if (shard == std::nullopt) {
+        vlog(clusterlog.warn, "can't find a shard for {}", ntp);
```
I think so, it isn't something which happens often and when it does it's an indication of an error

rystsov commented on src/v/cluster/tx_gateway.cc:
```diff
@@ -38,23 +38,31 @@ tx_gateway::init_tm_tx(init_tm_tx_request&&, rpc::streaming_context&) {
-tx_gateway::begin_tx(begin_tx_request&&, rpc::streaming_context&) {
-    return ss::make_ready_future<begin_tx_reply>(begin_tx_reply());
+tx_gateway::begin_tx(begin_tx_request&& request, rpc::streaming_context&) {
+    return _rm_partition_frontend.local().do_begin_tx(request.ntp, request.pid);
```
https://github.com/vectorizedio/redpanda/pull/1314#discussion_r624612252 also added a comment about it and an explicit leadership check (I didn't have it in the first place because it's anyway maybe become stale after the request execution started so we need to tolerate this)

rystsov commented on src/v/config/configuration.cc:
```diff
@@ -246,14 +246,14 @@ configuration::configuration()
       "Delay before retry a topic lookup in a shard or other meta tables",
       required::no,
-      0'100ms)
+      0'500ms)
```
I've made them abortable so responsiveness shouldn't be a problem

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+    }
+
+    if (!shard) {
+        vlog(clusterlog.warn, "can't find a shard for {}", model::kafka_tx_ntp);
```
should I do something like `vlog(clusterlog.warn, fmt_with_ctx(ssx::sformat, "can't find a shard for {}", model::kafka_tx_ntp));` ?

dotnwat commented on src/v/cluster/tx_gateway.cc:
```diff
@@ -56,24 +55,35 @@ tx_gateway::abort_tx(abort_tx_request&&, rpc::streaming_context&) {
-  commit_group_tx_request&&, rpc::streaming_context&) {
-    co_return commit_group_tx_reply();
+  commit_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_proxy->commit_group_tx_locally(
```
are there safety mechanisms in place in the `*_locally(...)` variants that deal with race conditions where an operation is dispatched over the network, but the state of the system has changed in the mean time such that the receiving node is no longer the correct node to handle the operation?

rystsov replied:
this concern is always valid because a system may change leadership after the check has passed. here in the transactional methods of group.cc we control it by avoiding mutating important state before replication and when we do conditional replication to avoid ABA problem

dotnwat commented on src/v/cluster/partition.h:
```diff
@@ -148,7 +148,7 @@ class partition {
         return _raft->get_configuration_manager();
     }
 
-    ss::shared_ptr<cluster::rm_stm>& rm_stm() { return _rm_stm; }
```
Noah's review placeholder

dotnwat reviewed (commented):
looks good. just adding this partial review I'm about half way through.

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -241,22 +287,24 @@ static ss::future<> do_fill_fetch_responses(
+                 hw = res.high_watermark,
+                 lso = res.last_stable_offset,
+                 pid = res.partition,
+                 rres = &res,
```
nit: I think you can do `&rres = res` to capture the reference rather than a raw pointer. alternatively, capture `aborted{std::move(res.aborted_transactions}}`

dotnwat commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -554,7 +604,7 @@ void op_context::create_response_placeholders() {
                 .error_code = error_code::none,
                 .high_watermark = fp.high_watermark,
-                .last_stable_offset = fp.high_watermark,
+                .last_stable_offset = fp.last_stable_offset,
```
is this fixing a bug unrelated to supporting transactions? if so, please note it in the commit message.

dotnwat reviewed (commented):
excellent pr. majority of my comments are optional _nit:_'s. there are a couple questions that seem important, and a couple spots that look like minor lifetime bugs that are easy to fix (if they are actually problems).

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+ss::future<> tx_gateway_frontend::stop() { return _gate.close(); }
+
+ss::future<std::optional<model::node_id>>
+tx_gateway_frontend::get_tx_broker([[maybe_unused]] ss::sstring key) {
```
was going to use it later, removing for now

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+             timeout,
+             outcome = std::move(outcome)]() mutable {
+                if (request.committed) {
+                    (void)self.commit_tm_tx(
```
right, it's an oversight fixing

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+  model::timeout_clock::duration timeout,
+  ss::promise<tx_errc>&& o) {
+    // moving to "stack" to be coroutine friendly
+    ss::promise<tx_errc> outcome = std::move(o);
```
can't do it - promise has removed copy constructor

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+  model::timeout_clock::duration timeout,
+  ss::promise<tx_errc>&& o) {
+    // moving to "stack" to be coroutine friendly
+    ss::promise<tx_errc> outcome = std::move(o);
```
can't do it - promise has removed copy constructor

dotnwat replied:
that's ok, you can still std::move from it and move to it. is there something else that's preventing it?

rystsov replied:
i was sure that I tried it before.. but it works now, changed

rystsov commented on src/v/cluster/types.h:
```diff
@@ -104,8 +106,57 @@ inline std::error_code make_error_code(tx_errc e) noexcept {
+    bool committed{};
+};
+struct end_tx_reply {
+    std::chrono::milliseconds throttle_time_ms{0};
```
don't know how it ended up there, removing

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+
+          (void)ss::with_gate(
+            self._gate,
+            [request = std::move(request),
```
fixed

rystsov commented on src/v/kafka/server/handlers/init_producer_id.cc:
```diff
@@ -31,6 +32,14 @@ ss::future<response_ptr> init_producer_id_handler::handle(
+            init_producer_id_response reply;
+            reply.data.error_code = error_code::cluster_authorization_failed;
+            return ctx.respond(reply);
+        }
```
right, when it isn't set clients use init_producer_id for idempotent producing which we also need to validate

dotnwat replied:
sorry, what i wasn't being very precise. are you sure it should _always_ be checked? the kafka code seems to only check it in cases where the transaction-id is null. but this change appears to check it unconditionally.

https://github.com/apache/kafka/blob/2.6/core/src/main/scala/kafka/server/KafkaApis.scala#L1943

```
    if (transactionalId != null) {
      if (!authorize(request.context, WRITE, TRANSACTIONAL_ID, transactionalId)) {
        sendErrorResponseMaybeThrottle(request, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED.exception)
        return
      }
    } else if (!authorize(request.context, IDEMPOTENT_WRITE, CLUSTER, CLUSTER_NAME)) {
      sendErrorResponseMaybeThrottle(request, Errors.CLUSTER_AUTHORIZATION_FAILED.exception)
      return
    }
```

rystsov replied:
fixed and created an issue to follow up - https://github.com/vectorizedio/redpanda/issues/1366

rystsov commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -209,17 +226,46 @@ ss::future<read_result> read_from_ntp(
     auto max_offset = high_watermark < model::offset(0) ? model::offset(0)
                                                         : high_watermark;
+
+    if (config::shard_local_cfg().enable_transactions.value()) {
```
it's easier this way because request context lives on another core and a couple of frames up the call stack, what do you think? I probaly could pass `are_transactions_enabled` bool although

dotnwat replied:
ahh, yeh getting at that state is pretty inconvenient. I think its fine to deal with this later.

rystsov commented on src/v/cluster/rm_partition_frontend.cc:
```diff
@@ -44,4 +43,444 @@ rm_partition_frontend::rm_partition_frontend(
+                commit_tx_reply{.ec = tx_errc::stm_not_found});
+          }
+
+          return stm->commit_tx(pid, tx_seq, timeout).then([](tx_errc ec) {
```
stm's type is `shared<cluster::tm_stm>`, its "master" copy is a member of partition so it outlives the current method and as I understand livetime isn't a problem here

rystsov replied:
https://github.com/vectorizedio/redpanda/issues/1365

rystsov commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -42,6 +71,884 @@ tx_gateway_frontend::tx_gateway_frontend(
+                  request, tx_errc::unknown_server_error));
+          }
+
+          return stm->get_end_lock(request.transactional_id)
```
https://github.com/vectorizedio/redpanda/pull/1314#discussion_r628852848

rystsov commented on src/v/cluster/partition.h:
```diff
@@ -168,6 +184,7 @@ class partition {
     ss::lw_shared_ptr<raft::log_eviction_stm> _nop_stm;
     ss::lw_shared_ptr<cluster::id_allocator_stm> _id_allocator_stm;
     ss::shared_ptr<cluster::rm_stm> _rm_stm;
+    ss::shared_ptr<cluster::tm_stm> _tm_stm;
```
slitted and merged with another commit

dotnwat commented on src/v/cluster/types.h:
```diff
@@ -104,6 +106,94 @@ inline std::error_code make_error_code(tx_errc e) noexcept {
+    kafka::group_id group_id{};
+};
+struct add_offsets_tx_reply {
+    std::chrono::milliseconds throttle_time_ms{0};
```
throttle_time?

dotnwat commented on src/v/cluster/types.h:
```diff
@@ -104,6 +106,94 @@ inline std::error_code make_error_code(tx_errc e) noexcept {
+        model::topic name{};
+        std::vector<add_paritions_tx_reply::partition_result> results{};
+    };
+    std::chrono::milliseconds throttle_time_ms{0};
```
throttle_time?

dotnwat reviewed (commented):
revisions look great! I think we missed a couple throttle time fields that snuck into the internal rpc reply types. From the previous review revisions it looks like this was the only one that was unresolved: https://github.com/vectorizedio/redpanda/pull/1314#discussion_r628901571

dotnwat commented:
manually restarted CI after two raft unit test failures. we've been seeing those periodically recently.

senior7515 commented:
wow! congratulations @rystsov and thank you @dotnwat for reviewing diligently. 🎉 🙌🏽 🔥 🚀 🌀
